examples/Ex02_CustomTypeProviders/Ex02_CustomTypeProvidersTest.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+// This file is part of Phodam
+// Copyright (c) Andrew Vehlies <avehlies@gmail.com>
+// Licensed under the MIT license. See LICENSE file in the project root.
+// SPDX-License-Identifier: MIT
+
+declare(strict_types=1);
+
+namespace PhodamExamples\Ex02_CustomTypeProviders;
+
+use Phodam\Analyzer\TypeAnalysisException;
+use Phodam\Phodam;
+use Phodam\Provider\ProviderConfig;
+use PHPUnit\Framework\TestCase;
+
+class Ex02_CustomTypeProvidersTest extends TestCase
+{
+    private Phodam $phodam;
+
+    public function setUp(): void
+    {
+        parent::setUp();
+        $this->phodam = new Phodam();
+
+        // Tired of specifying active = true, 0.0 <= gpa <= 4.0 ?
+        // Make a provider to have your own defaults!
+        $studentProvider = new StudentTypeProvider();
+        $studentProviderConfig = (new ProviderConfig($studentProvider))
+            ->forType(Student::class);
+
+        $this->phodam->registerProviderConfig($studentProviderConfig);
+
+        // A Classroom provider is needed because `array $students`
+        // doesn't have a type, and we can't assume because of 'array'
+        $classroomProvider = new ClassroomTypeProvider();
+        $classroomProviderConfig = (new ProviderConfig($classroomProvider))
+            ->forType(Classroom::class);
+
+        $this->phodam->registerProviderConfig($classroomProviderConfig);
+    }
+
+    public function testBeforeRegistering(): void
+    {
+        // we expect a new instance of Phodam to not understand how to
+        // make a Classroom because it's unable to map `array $students` to
+        // a type it can generate
+        $localPhodam = new Phodam();
+        try {
+            $localPhodam->create(Classroom::class);
+        } catch (TypeAnalysisException $ex) {
+            $this->assertStringContainsString(
+                'PhodamExamples\Ex02_CustomTypeProviders\Classroom: Unable to map fields: students',
+                $ex->getMessage()
+            );
+        }
```

Comment:
It might be more idiomatic to use `expectException` and `expectExceptionMessage`.

```suggestion
        $localPhodam = new Phodam();

        $this->expectException(TypeAnalysisException:class);
        $this->expectExceptionMessage(
            'PhodamExamples\Ex02_CustomTypeProviders\Classroom: Unable to map fields: students'
        );

        $localPhodam->create(Classroom::class);
```

Reply:
Ah yup, this was a remnant from when the test was originally doing...

```php
try {
    $this->phodam->create(Classroom::class);
} catch (...) {}

// now register and generate
```

Reply:
8edb104 Resolved here